pkg/kubelet/kubelet.go

```diff
@@ -1839,6 +1839,17 @@ func (kl *Kubelet) syncLoopIteration(configCh <-chan kubetypes.PodUpdate, handle
 		kl.sourcesReady.AddSource(u.Source)
 
 	case e := <-plegCh:
+		if e.Type == pleg.ContainerStarted {
```

Comment:
talked offline... this doesn't solve the race observing a delete from configCh and oberving a ContainerStarted event from plegCh, but it does solve the issue where a container status reflected in the API (so the kubelet clearly knew the container was running) did not inform graceful deletion

My main question now is what the load impact is for a kubelet that is rapidly creating containers:

* does invalidating here cause any pathological issues for a pod with N containers (where each individual ContainerStarted event invalidates the cache and some other path calls GetPods() per-container, resulting in N cache misses.
* do we only care that GetPods() is super up-to-date in the delete case? would it make sense to use the ForceUpdateIfOlder() method in deletePod() passing in the max status.{init}containers[*].state.running.startedAt or something?

Suggestions if we keep the current approach:

1. a test exercising this would be great if it's possible
2. I'm not sure if it would be possible to move the invalidate to the point where we make the call to create the container or not

/cc @dashpole for kubelet sync loop review expertise as well

Reply:
> * do we only care that GetPods() is super up-to-date in the delete case? would it make sense to use the ForceUpdateIfOlder() method in deletePod()?

a sketch of that approach is at https://github.com/liggitt/kubernetes/commits/force-delete ... that would let us keep the benefits of the runtimeCache in most cases (only rapid start/delete cases would result in cache misses)

Reply:
I like the deletion-only approach.  The only thing I need to check is IIRC there are a few different deletion paths.  We should make sure we actually cover all deletion cases...

Reply:
Looks like we just need to add the update logic in one other location